### PR TITLE
[Mentions v2] Parse <mention-report> when displaying reportAction

### PR DESCRIPTION
**Pull request by war-in**
### Fixed Issues

$ https://github.com/Expensify/App/issues/39545
PROPOSAL:

### Tests

To check the functionality paste the code in `src/pages/home/report/ReportActionsView.tsx` under the `PopoverReactionList` (line 533)
```
<RenderHTML html="<comment><mention-report>#admins</mention-report></comment>" />
<RenderHTML html='<comment><mention-report reportID="3088255940953979" /></comment>' />
<RenderHTML html="<comment><mention-report>#non-existent-room</mention-report></comment>" />
<RenderHTML html='<comment><mention-report reportID="111111" /></comment>' />        
```

1. Go to a chat
2. There should be some tags
image
3. The first one is clickable and redirects to the `#admins` room
4. The second one is clickable and redirects to the chat of id `3088255940953979` (you probably would need to change the id to one existing on your account)
5. Last two should not redirect because there are no such rooms

### QA Steps

To check the functionality paste the code in `src/pages/home/report/ReportActionsView.tsx` under the `PopoverReactionList` (line 533)
```
<RenderHTML html="<comment><mention-report>#admins</mention-report></comment>" />
<RenderHTML html='<comment><mention-report reportID="3088255940953979" /></comment>' />
<RenderHTML html="<comment><mention-report>#non-existent-room</mention-report></comment>" />
<RenderHTML html='<comment><mention-report reportID="111111" /></comment>' />        
```

1. Go to a chat
2. There should be some tags
image
3. The first one is clickable and redirects to the `#admins` room
4. The second one is clickable and redirects to the chat of id `3088255940953979` (you probably would need to change the id to one existing on your account)
5. Last two should not redirect because there are no such rooms

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x ] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/61014013/8dd50e1e-e202-4f4b-9c3a-d09d45aa93e2

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/61014013/2be58385-e640-4782-a937-bca059c693e0

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/61014013/f536275a-30f1-46bf-a370-f108beee77d6

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/61014013/9d96f9c6-73e3-482a-a6f2-842bdb2475a8

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/61014013/a50d071e-cbf1-4fa9-ae20-a48f8b5515d6

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/61014013/626c5fc1-c477-4e8b-b095-71bacadf4603

</details>

### CODE REVIEW

File: src/components/HTMLEngineProvider/HTMLRenderers/MentionRoomRenderer.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+function MentionRoomRenderer({style, tnode, TDefaultRenderer, reports, ...defaultRendererProps}: MentionRoomRendererProps) {
+    const styles = useThemeStyles();
+    const StyleUtils = useStyleUtils();
+    const htmlAttributeReportID = tnode.attributes.reportid;
+    const currentReportID = useCurrentReportID();
+
+    let reportID: number | undefined;
+    let mentionDisplayText: string;
+
+    const removeLeadingLTRAndHash = (value: string) => value.replace(CONST.UNICODE.LTR, '').slice(1);
```
Comment (robertKozik) [inline src/components/HTMLEngineProvider/HTMLRenderers/MentionRoomRenderer.tsx:37]: get it outside the component

File: src/components/HTMLEngineProvider/HTMLRenderers/MentionRoomRenderer.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        // eslint-disable-next-line rulesdir/prefer-early-return
+        Object.values(reports ?? {}).forEach((report) => {
+            if (report?.policyID === currentReport?.policyID && removeLeadingLTRAndHash(report?.reportName ?? '') === mentionDisplayText) {
+                reportID = Number(report?.reportID);
+            }
+        });
+    } else {
+        return null;
+    }
```
Comment (robertKozik) [inline src/components/HTMLEngineProvider/HTMLRenderers/MentionRoomRenderer.tsx:57]: extract to separate function

File: src/components/HTMLEngineProvider/HTMLRenderers/MentionRoomRenderer.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    let reportID: number | undefined;
+    let mentionDisplayText: string;
+
+    const removeLeadingLTRAndHash = (value: string) => value.replace(CONST.UNICODE.LTR, '').slice(1);
+
+    if (!isEmpty(htmlAttributeReportID)) {
+        const report = reports?.['report_'.concat(htmlAttributeReportID)];
+
+        reportID = report?.reportID ? parseInt(report.reportID, 10) : undefined;
```
Comment (robertKozik) [inline src/components/HTMLEngineProvider/HTMLRenderers/MentionRoomRenderer.tsx:42]: remove `parseInt`

File: src/components/HTMLEngineProvider/HTMLRenderers/MentionRoomRenderer.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        </ShowContextMenuContext.Consumer>
+    );
+}
+
+MentionRoomRenderer.displayName = 'MentionRoomRenderer';
+
+export default withOnyx<MentionRoomRendererProps, RoomMentionOnyxProps>({
+    reports: {
+        key: ONYXKEYS.COLLECTION.REPORT,
+    },
```
Comment (robertKozik) [inline src/components/HTMLEngineProvider/HTMLRenderers/MentionRoomRenderer.tsx:95]: limit reports to only to used values

**Comment by robertKozik** [comment]
Hi! We were in the internal review process, but the PR is already available for review 🚀

**Comment by puneetlath** [comment]
I think we should call the renderer `MentionReport` instead of `MentionRoom` since the tag is `<mention-report>`

**Comment by robertKozik** [comment]
> Linking to another room only seems to work from rooms, not from a normal chat. `navigationRoute` seems to be undefined when attempting to navigate from a normal chat:

Actually @rlinoz @puneetlath When user types by hand room mention (auto complete is disabled for chats which are not in group policy) we will parse it and send to backend (parser is not aware of chat type). Will backend filtered this tag out?

**Comment by robertKozik** [comment]
> I think we should call the renderer `MentionReport` instead of `MentionRoom` since the tag is `<mention-report>`

Changed 👌🏼

**Comment by robertKozik** [comment]
So if we have only optimistic version of the report action the expected behaviour would be to check if mentions are allowed in current report and if not just return null and not render it at all?

**Comment by rlinoz** [comment]
I think we should strip out the tag and render as plain text if the mention is happening in a report that is not allowed.

**Comment by rlinoz** [comment]
> I think we should strip out the tag and render as plain text if the mention is happening in a report that is not allowed.

Just to be sure, we are going to do that in this PR right?

**Comment by robertKozik** [comment]
Yeah @war-in will take care of it tommorow

File: src/components/HTMLEngineProvider/HTMLRenderers/MentionReportRenderer.tsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+const removeLeadingLTRAndHash = (value: string) => value.replace(CONST.UNICODE.LTR, '').slice(1);
+
+const getMentionDetails = (htmlAttributeReportID: string, currentReportID: string, reports: OnyxCollection<Report>, tnode: TText | TPhrasing) => {
+    let reportID: string | undefined;
+    let mentionDisplayText: string;
+
+    // get mention details based on reportID from tag attribute
+    if (!isEmpty(htmlAttributeReportID)) {
+        const report = reports?.['report_'.concat(htmlAttributeReportID)];
```
Comment (robertKozik) [inline src/components/HTMLEngineProvider/HTMLRenderers/MentionReportRenderer.tsx:35]:
```suggestion
        const report = getReport(htmlAttributeReportID)
```

File: src/components/HTMLEngineProvider/HTMLRenderers/MentionReportRenderer.tsx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    return {reportID, mentionDisplayText};
+};
+
+function MentionReportRenderer({style, tnode, TDefaultRenderer, reports, ...defaultRendererProps}: MentionRoomRendererProps) {
+    const styles = useThemeStyles();
+    const StyleUtils = useStyleUtils();
+    const htmlAttributeReportID = tnode.attributes.reportid;
+    const currentReportID = useCurrentReportID();
```
Comment (robertKozik) [inline src/components/HTMLEngineProvider/HTMLRenderers/MentionReportRenderer.tsx:61]: You can destructure here
```suggestion
    const {currentReportID} = useCurrentReportID();
```

**Comment by war-in** [comment]
@Ollyws fixed! Now, linking from the chat is disabled as @rlinoz requested

**Review by Ollyws — APPROVED** [review]
LGTM.

File: src/components/HTMLEngineProvider/HTMLRenderers/MentionReportRenderer.tsx [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+                            ? (event) => {
+                                  event.preventDefault();
+                                  Navigation.navigate(navigationRoute);
+                              }
+                            : undefined
+                    }
+                    role={isGroupPolicyReport ? CONST.ROLE.LINK : undefined}
+                    accessibilityLabel={isGroupPolicyReport ? `/${navigationRoute}` : undefined}
+                >
+                    #{mentionDisplayText}
```
Comment (rlinoz) [inline src/components/HTMLEngineProvider/HTMLRenderers/MentionReportRenderer.tsx:103]: I think this is duplicating the `#` for `reportID` mentions.
image

File: src/components/HTMLEngineProvider/HTMLRenderers/MentionReportRenderer.tsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+type MentionReportRendererProps = MentionReportOnyxProps & CustomRendererProps<TText | TPhrasing>;
+
+const removeLeadingLTRAndHash = (value: string) => value.replace(CONST.UNICODE.LTR, '').slice(1);
+
+const getMentionDetails = (htmlAttributeReportID: string, currentReport: OnyxEntry<Report> | EmptyObject, reports: OnyxCollection<Report>, tnode: TText | TPhrasing) => {
+    let reportID: string | undefined;
+    let mentionDisplayText: string;
+
+    // get mention details based on reportID from tag attribute
```
Comment (rlinoz) [inline src/components/HTMLEngineProvider/HTMLRenderers/MentionReportRenderer.tsx:36]:
```suggestion
    // Get mention details based on reportID from tag attribute
```

File: src/components/HTMLEngineProvider/HTMLRenderers/MentionReportRenderer.tsx [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    if (!isEmpty(htmlAttributeReportID)) {
+        const report = getReport(htmlAttributeReportID);
+
+        reportID = report?.reportID ?? undefined;
+        mentionDisplayText = report?.reportName ?? report?.displayName ?? htmlAttributeReportID;
+        // get mention details from name inside tnode
```
Comment (rlinoz) [inline src/components/HTMLEngineProvider/HTMLRenderers/MentionReportRenderer.tsx:42]:
```suggestion
        // Get mention details from name inside tnode
```

**Review by rlinoz — CHANGES_REQUESTED** [review]

**Comment by war-in** [comment]
@rlinoz fixed 😌

**Review by puneetlath — APPROVED** [review]

**Review by rlinoz — APPROVED** [review]
Thanks!

**Comment by kbecciv** [comment]
@war-in Can you verify internally, not sure if QA team can paste the code in src/pages/home/report/ReportActionsView.tsx under the PopoverReactionList (line 533)?

**Comment by robertKozik** [comment]
As I saw that other PR's related to mentions are deployed as well, you should be able to just type the mention inside composer - sent message should contain the mention and be styled properly.

**Pull request merged**